### PR TITLE
magento/magento2: Fixes for the schema cache.xsd

### LINKED ISSUE #26224: Cache type without "instance" causes exception when disabling the module through "Cache Management" in the backend
### Preconditions (*)
1. Magento 2.3.3 CE

### Steps to reproduce (*)

1. Create a cache type using "cache.xml" without the "instance" argument
```xml
<config .../>
    <type name="exceptional_cache_type">
        <label>Exceptional Cache Type</label>
        <description>An exceptional cache type using a quantum computer</description>
    </type>
</config>
```
2. Go to "System > Cache Management" in the backend
3. Check the checkbox of the cache type "Exceptional Cache Type"
4. Select "Disable" and click "Submit"

### Expected result (*)
I'd expect the cache type to be successfully disabled. 

According to the XSD ("lib/internal/Magento/Framework/Cache/etc/cache.xsd") the "instance" argument is optional and not required.

### Actual result (*)
It throws an exception because it's trying to invoke `->clean(...)` on a null variable. 

```
Fatal error: Uncaught Error: Call to a member function clean() on null in /lib/internal/Magento/Framework/App/Cache/TypeList.php:194 Stack trace:
#0 /app/code/Magento/Backend/Controller/Adminhtml/Cache/MassDisable.php(62): Magento\Framework\App\Cache\TypeList->cleanType('invalid_cache_t...')
#1 /app/code/Magento/Backend/Controller/Adminhtml/Cache/MassDisable.php(41): Magento\Backend\Controller\Adminhtml\Cache\MassDisable->disableCache()
#2 /generated/code/Magento/Backend/Controller/Adminhtml/Cache/MassDisable/Interceptor.php(24): Magento\Backend\Controller\Adminhtml\Cache\MassDisable->execute()
#3 /lib/internal/Magento/Framework/App/Action/Action.php(107): Magento\Backend\Controller\Adminhtml\Cache\MassDisable\Interceptor->execute()
```

### PR DESCRIPTION
**Pull request by andrewbess**
### Description (*)

This PR fixes the schema "cache.xsd" from Magento\Framework\Cache.

1. The attribute "instance" has changed on required because if a developer won't be to use it new cache type will not be created;
2. The parameters "label" and "description" have changed on required because of the class "Magento\Framework\App\Cache\TypeList" gets theirs without checking for existing. Please take a look at [this code](https://github.com/magento/magento2/blob/cde4f67d8e02ad7ffe1de5051010bdce71fdabfd/lib/internal/Magento/Framework/App/Cache/TypeList.php#L124);
3. Checking on unique of the type name has added;
4. The configuration reader class has fixed for more informative during to run CLI commands;
5. The class SchemaLocator has been fixed to check the cache configuration during configuration merging;
6. The unnecessary checking has removed from the class TypeList;
7. Unit tests have been covered.

### Fixed Issues (if relevant)

Fixes: https://github.com/magento/magento2/issues/26224

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by lbajsarowicz** [comment]
Just add `WIP:` prefix to your PR name.

**Review by swnsma — APPROVED** [review]
Failed function tests do not related to the improved functionality.

**Comment by magento-engcom-team** [comment]
Hi @swnsma, thank you for the review.
[ENGCOM-7114](https://jira.corp.magento.com/browse/ENGCOM-7114) has been created to process this Pull Request
:eight_spoked_asterisk: @swnsma, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Comment by lbajsarowicz** [comment]
@swnsma It's not finished yet? :-)

**Comment by swnsma** [comment]
@lbajsarowicz endless story :)

**Comment by andrewbess** [comment]
Hello @lbajsarowicz 
It has already finished.

File: dev/tests/static/testsuite/Magento/Test/Integrity/Magento/Framework/Cache/ConfigTest.php [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+ * See COPYING.txt for license details.
+ */
+declare(strict_types = 1);
+
+namespace Magento\Test\Integrity\Magento\Framework\Cache;
+
+/**
+ * Unit test of the cache configuration
+ */
+class ConfigTest extends \PHPUnit\Framework\TestCase
```
Comment (lbajsarowicz) [inline dev/tests/static/testsuite/Magento/Test/Integrity/Magento/Framework/Cache/ConfigTest.php]: You can definately import the classes :-)

File: dev/tests/static/testsuite/Magento/Test/Integrity/Magento/Framework/Cache/ConfigTest.php [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+        $this->assertEmpty($actualResult);
+    }
+
+    /**
+     * Data provider with invalid xml array according to cache.xsd
+     */
+    public function schemaCorrectlyIdentifiesInvalidXmlDataProvider(): array
+    {
+        return include __DIR__ . '/_files/invalidCacheConfigXmlArray.php';
```
Comment (lbajsarowicz) [inline dev/tests/static/testsuite/Magento/Test/Integrity/Magento/Framework/Cache/ConfigTest.php:87]: Interesting approach to provide code samples.

File: lib/internal/Magento/Framework/App/Test/Unit/Cache/TypeListTest.php [hunk 3]
```diff
@@ -6,21 +6,30 @@
...
 {
     /**
-     * @var \Magento\Framework\App\Cache\TypeList
+     * @var TypeList
      */
     protected $_typeList;
 
     /**
-     * @var \Magento\Framework\App\CacheInterface|\PHPUnit_Framework_MockObject_MockObject
+     * @var CacheInterface|PHPUnit_Framework_MockObject_MockObject
```
Comment (lbajsarowicz) [inline lib/internal/Magento/Framework/App/Test/Unit/Cache/TypeListTest.php]: Replace with `\PHPUnit\Framework\MockObject\MockObject`

File: lib/internal/Magento/Framework/Cache/Config/Reader.php [hunk 4]
```diff
@@ -24,7 +30,7 @@ class Reader extends \Magento\Framework\Config\Reader\Filesystem
      * @param string $domDocumentClass
      * @param string $defaultScope
      */
-    public function __construct(
+    public function __construct( //phpcs:ignore Generic.CodeAnalysis.UselessOverridingMethod
```
Comment (lbajsarowicz) [inline lib/internal/Magento/Framework/Cache/Config/Reader.php]: Move that to DocBlock above.

File: lib/internal/Magento/Framework/Cache/Config/Reader.php [hunk 5]
```diff
@@ -24,7 +30,7 @@ class Reader extends \Magento\Framework\Config\Reader\Filesystem
...
         \Magento\Framework\Config\FileResolverInterface $fileResolver,
         \Magento\Framework\Cache\Config\Converter $converter,
         \Magento\Framework\Cache\Config\SchemaLocator $schemaLocator,
```
Comment (lbajsarowicz) [inline lib/internal/Magento/Framework/Cache/Config/Reader.php]: Import classes

**Review by lbajsarowicz — CHANGES_REQUESTED** [review]
Requesting changes due to use of deprecated `MockObject`. Please replace it with proper one :-)

File: dev/tests/static/testsuite/Magento/Test/Integrity/Magento/Framework/Cache/ConfigTest.php (hunk 1 above)
Comment (andrewbess) [reply dev/tests/static/testsuite/Magento/Test/Integrity/Magento/Framework/Cache/ConfigTest.php]: Hello @lbajsarowicz 
Thank you for your remark.
I have fixed it. Please check.
Thank you in advance.

File: lib/internal/Magento/Framework/App/Test/Unit/Cache/TypeListTest.php (hunk 3 above)
Comment (andrewbess) [reply lib/internal/Magento/Framework/App/Test/Unit/Cache/TypeListTest.php]: Hello @lbajsarowicz 
Thank you for your remark.
I have fixed it. Please check.
Thank you in advance.

File: lib/internal/Magento/Framework/Cache/Config/Reader.php (hunk 4 above)
Comment (andrewbess) [reply lib/internal/Magento/Framework/Cache/Config/Reader.php]: Hello @lbajsarowicz 
Thank you for your remark.
I have fixed it. Please check.
Thank you in advance.

File: lib/internal/Magento/Framework/Cache/Config/Reader.php (hunk 5 above)
Comment (andrewbess) [reply lib/internal/Magento/Framework/Cache/Config/Reader.php]: Hello @lbajsarowicz 
Thank you for your remark.
I have fixed it. Please check.
Thank you in advance.

File: dev/tests/static/testsuite/Magento/Test/Integrity/Magento/Framework/Cache/ConfigTest.php (hunk 2 above)
Comment (andrewbess) [reply dev/tests/static/testsuite/Magento/Test/Integrity/Magento/Framework/Cache/ConfigTest.php:87]: Hello @lbajsarowicz 
Thank you for your remark.
Magento already has the same data providers. Please take a look at the screenshot
interesting-data-providers
Thank you in advance.

**Review by lbajsarowicz — CHANGES_REQUESTED** [review]
Please fix **Static Tests** and I'm ready to :heavy_check_mark:

**Comment by andrewbess** [comment]
Hello @lbajsarowicz 
I fixed **Static Tests** 
Please check.
Thank you in advance.

**Review by lbajsarowicz — APPROVED** [review]
:heavy_check_mark: Thank you for your contribution

**Comment by magento-engcom-team** [comment]
Hi @lbajsarowicz, thank you for the review.
[ENGCOM-7114](https://jira.corp.magento.com/browse/ENGCOM-7114) has been created to process this Pull Request
:eight_spoked_asterisk: @lbajsarowicz, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Comment by andrewbess** [comment]
Hello @lbajsarowicz 
Perhaps, current PR is covered tests.
I have made the new test and fixed the existing test.

**Comment by swnsma** [comment]
Integrity tests were introduced in this PR to check updated XSD schema.

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

**Before:**
before

**After:**
after

**Comment by engcom-Echo** [comment]
Failed static tests not related to the changes in this PR

**Pull request merged**